ngmix/ksigmamom.py

```diff
@@ -0,0 +1,445 @@
+        # like this
+        #
+        #  deconv_im = kim / kpsf_im * (im_cen_phase / psf_im_cen_phase)
+        #
+        # The phases are complex exponentials
+        #
+        #  exp(ik*cen)
+        #
+        # So we can compute one phase as
+        #
+        #  im_cen_phase / psf_im_cen_phase = exp(ik * (im_cen - psf_cen))
+        #
+        # and then multiply it into the image.
+        #
+        # This operation and the deconvolutiomn will be done
+        # later in _measure_moments_fft
+
+        # now build the kernels
+        kres = _ksigma_kernels(
+            target_dim,
+            self.fwhm,
+            obs.jacobian.dvdrow, obs.jacobian.dvdcol,
+            obs.jacobian.dudrow, obs.jacobian.dudcol,
+        )
+
+        # compute the total variance from weight map
+        msk = obs.weight > 0
+        tot_var = np.sum(1.0 / obs.weight[msk])
+
+        # run the actual measurements and return
+        res = _measure_moments_fft(
+            kim, kpsf_im, tot_var, eff_pad_factor, kres,
+            im_row - psf_im_row, im_col - psf_im_col,
+        )
+        if res['flags'] != 0:
+            logger.debug("ksigma pre-psf moments failed: %s" % res['flagstr'])
+
+        if return_kernels:
+            # put the kernels back into their unpacked state
+            new_kres = {}
+            for k in kres:
+                if k == "msk":
+                    continue
+                new_kres[k] = np.zeros((fft_dim, fft_dim), dtype=np.complex128)
+                new_kres[k][kres["msk"]] = kres[k]
+            res["kernels"] = new_kres
+
+        return res
+
+
+def _measure_moments_fft(kim, kpsf_im, tot_var, eff_pad_factor, kernels, drow, dcol):
+    # we only need to do things where the ksigma kernel is non-zero
+    # this saves a bunch of CPU cycles
+    msk = kernels["msk"]
+    dim = kim.shape[0]
+
+    # deconvolve PSF
+    kim, kpsf_im, _ = _deconvolve_im_psf_inplace(
+        kim[msk],
+        kpsf_im[msk],
+        # max amplitude is flux which is 0,0 in the standard FFT convention
+        np.abs(kpsf_im[0, 0]),
+    )
+
+    # put in phase shift as described above
+    # the sin and cos are expensive so we only compute them where we will
+    # use the image which is in the msk
+    if drow != 0 or dcol != 0:
+        cen_phase = _compute_cen_phase_shift(drow, dcol, dim, msk=msk)
+        kim *= cen_phase
+
+    # build the flux, radial, plus and cross kernels / moments
+    # the inverse FFT in our convention has a factor of 1/n per dimension
+    # the sums below are inverse FFTs but only computing the values at the
+    # real-space center of the object (0 in our coordinate system).
+    # thus we code the factor of 1/n by hand
+    df = 1/dim
+    df2 = df * df
+    df4 = df2 * df2
+
+    # we only sum where the kernel is nonzero
+    fkf = kernels["fkf"]
+    fkr = kernels["fkr"]
+    fkp = kernels["fkp"]
+    fkc = kernels["fkc"]
+
+    mf = np.sum((kim * fkf).real) * df2
```

Comment:
bunch of temporary arrays here, inefficient.  Have you profiled to see where the most time is spent?  Is it the ffts?

Reply:
Yes. Most of the time is spent doing FFTs or generating the kernels themselves.